File: go/format.go
package main

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
)

// Wire format v3. See ../SPEC.md — the Android app parses the exact same bytes.
const (
	magic   = "ZWS3"
	version = 3

	flagEncrypted  = 1 << 0
	flagCompressed = 1 << 1

	kdfNone         = 0
	kdfPBKDF2SHA256 = 1

	saltLen  = 16
	nonceLen = 12

	headerFixed = 8  // magic + version + flags + kdf + reserved
	lenField    = 4  // payloadLen
	padBlock    = 64 // inner frame is padded to a multiple of this when encrypted

	// maxPayload bounds allocations when parsing hostile input.
	maxPayload = 16 << 20
)

var (
	errNoPayload  = errors.New("no hidden data found in this text")
	errBadMagic   = errors.New("invisible characters were found, but they are not a Sayeh payload (truncated, reformatted, or produced by another tool)")
	errTruncated  = errors.New("the payload is truncated — the text was cut off or reformatted in transit")
	errBadVersion = errors.New("this payload was made by a newer version of Sayeh")
)

// container is the parsed, still-sealed form of a payload.
type container struct {
	flags      byte
	kdf        byte
	iterations uint32
	salt       []byte
	nonce      []byte
	payload    []byte
}

func (c *container) encrypted() bool  { return c.flags&flagEncrypted != 0 }
func (c *container) compressed() bool { return c.flags&flagCompressed != 0 }

// header returns every byte that precedes the payload, including the length
// field. It doubles as the GCM additional authenticated data, which is what
// binds the KDF parameters and the declared length to the ciphertext.
func (c *container) header(payloadLen int) []byte {
	buf := make([]byte, 0, headerFixed+4+saltLen+nonceLen+lenField)
	buf = append(buf, magic...)
	buf = append(buf, version, c.flags, c.kdf, 0)

	if c.encrypted() {
		buf = binary.BigEndian.AppendUint32(buf, c.iterations)
		buf = append(buf, c.salt...)
		buf = append(buf, c.nonce...)
	}
	return binary.BigEndian.AppendUint32(buf, uint32(payloadLen))
}

// marshal serializes header + payload.
func (c *container) marshal() []byte {
	return append(c.header(len(c.payload)), c.payload...)
}

// parseContainer validates and splits raw bytes recovered from the carriers.
// It returns the container plus the header slice used as AAD.
func parseContainer(raw []byte) (*container, []byte, error) {
	if len(raw) < headerFixed+lenField {
		if len(raw) == 0 {
			return nil, nil, errNoPayload
		}
		return nil, nil, errBadMagic
	}
	if !bytes.HasPrefix(raw, []byte(magic)) {
		return nil, nil, errBadMagic
	}
	if raw[4] != version {
		return nil, nil, errBadVersion
	}

	c := &container{flags: raw[5], kdf: raw[6]}
	off := headerFixed

	if c.encrypted() {
		need := off + 4 + saltLen + nonceLen + lenField
		if len(raw) < need {
			return nil, nil, errTruncated
		}
		if c.kdf != kdfPBKDF2SHA256 {
			return nil, nil, fmt.Errorf("unsupported key derivation function id %d", c.kdf)
		}
		c.iterations = binary.BigEndian.Uint32(raw[off:])
		off += 4
		c.salt = raw[off : off+saltLen]
		off += saltLen
		c.nonce = raw[off : off+nonceLen]
		off += nonceLen

		if c.iterations < minIterations {
			return nil, nil, fmt.Errorf("refusing payload: iteration count %d is below the safe minimum of %d", c.iterations, minIterations)
		}
	} else if c.kdf != kdfNone {
		return nil, nil, errBadMagic
	}

	if len(raw) < off+lenField {
		return nil, nil, errTruncated
	}
	n := binary.BigEndian.Uint32(raw[off:])
	off += lenField

	if n > maxPayload {
		return nil, nil, fmt.Errorf("declared payload size (%d bytes) exceeds the %d MiB limit", n, maxPayload>>20)
	}
	if uint64(len(raw)-off) < uint64(n) {
		return nil, nil, errTruncated
	}

	header := raw[:off]
	c.payload = raw[off : off+int(n)]
	return c, header, nil
}

// packInner builds bodyLen || body || padding.
func packInner(body []byte, pad bool) ([]byte, error) {
	total := lenField + len(body)
	if pad {
		total = (total + padBlock - 1) / padBlock * padBlock
	}

	inner := make([]byte, total)
	binary.BigEndian.PutUint32(inner, uint32(len(body)))
	copy(inner[lenField:], body)

	if n := total - lenField - len(body); n > 0 {
		if err := randBytes(inner[lenField+len(body):]); err != nil {
			return nil, err
		}
	}
	return inner, nil
}

// unpackInner reverses packInner and drops the padding.
func unpackInner(inner []byte) ([]byte, error) {
	if len(inner) < lenField {
		return nil, errTruncated
	}
	n := binary.BigEndian.Uint32(inner)
	if uint64(n) > uint64(len(inner)-lenField) {
		return nil, errTruncated
	}
	return inner[lenField : lenField+int(n)], nil
}

File: legacy/v3/android/java/ir/sayeh/stego/Format.java
package ir.sayeh.stego;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.util.Arrays;

/** Wire format v3 ("ZWS3"). Mirrors format.go — see SPEC.md. */
public final class Format {

    public static final byte[] MAGIC = {'Z', 'W', 'S', '3'};
    public static final byte VERSION = 3;

    public static final int FLAG_ENCRYPTED = 1;
    public static final int FLAG_COMPRESSED = 2;

    public static final int KDF_NONE = 0;
    public static final int KDF_PBKDF2_SHA256 = 1;

    public static final int HEADER_FIXED = 8; // magic + version + flags + kdf + reserved
    public static final int LEN_FIELD = 4;
    public static final int PAD_BLOCK = 64;
    public static final int MAX_PAYLOAD = 16 << 20;

    private Format() {
    }

    /** A parsed, still-sealed payload. */
    public static final class Container {
        public int flags;
        public int kdf;
        public int iterations;
        public byte[] salt;
        public byte[] nonce;
        public byte[] payload;
        /** Every byte before the payload, including the length field. Used as GCM AAD. */
        public byte[] header;

        public boolean isEncrypted() {
            return (flags & FLAG_ENCRYPTED) != 0;
        }

        public boolean isCompressed() {
            return (flags & FLAG_COMPRESSED) != 0;
        }

        /**
         * Builds the header for a given payload length. It is authenticated as
         * AAD, so version, flags, KDF id, iteration count, salt, nonce and
         * length can none of them be edited without breaking the tag.
         */
        public byte[] buildHeader(int payloadLen) {
            ByteArrayOutputStream out = new ByteArrayOutputStream(64);
            out.write(MAGIC, 0, MAGIC.length);
            out.write(VERSION);
            out.write(flags);
            out.write(kdf);
            out.write(0); // reserved

            if (isEncrypted()) {
                writeInt(out, iterations);
                out.write(salt, 0, salt.length);
                out.write(nonce, 0, nonce.length);
            }
            writeInt(out, payloadLen);
            return out.toByteArray();
        }

        public byte[] marshal() {
            byte[] h = buildHeader(payload.length);
            byte[] all = new byte[h.length + payload.length];
            System.arraycopy(h, 0, all, 0, h.length);
            System.arraycopy(payload, 0, all, h.length, payload.length);
            return all;
        }
    }

    private static void writeInt(ByteArrayOutputStream out, int v) {
        out.write((v >>> 24) & 0xFF);
        out.write((v >>> 16) & 0xFF);
        out.write((v >>> 8) & 0xFF);
        out.write(v & 0xFF);
    }

    /** Validates and splits raw bytes recovered from the carriers. */
    public static Container parse(byte[] raw) throws StegoException {
        if (raw.length == 0) {
            throw StegoException.noPayload();
        }
        if (raw.length < HEADER_FIXED + LEN_FIELD || !startsWith(raw, MAGIC)) {
            throw StegoException.badMagic();
        }
        if (raw[4] != VERSION) {
            throw new StegoException(StegoException.Kind.BAD_VERSION,
                    "This payload was made by a newer version of Sayeh.");
        }

        Container c = new Container();
        c.flags = raw[5] & 0xFF;
        c.kdf = raw[6] & 0xFF;
        int off = HEADER_FIXED;

        if (c.isEncrypted()) {
            int need = off + 4 + Crypto.SALT_LEN + Crypto.NONCE_LEN + LEN_FIELD;
            if (raw.length < need) {
                throw StegoException.truncated();
            }
            if (c.kdf != KDF_PBKDF2_SHA256) {
                throw new StegoException(StegoException.Kind.CORRUPT,
                        "Unsupported key derivation function id " + c.kdf + ".");
            }
            c.iterations = ByteBuffer.wrap(raw, off, 4).getInt();
            off += 4;
            c.salt = Arrays.copyOfRange(raw, off, off + Crypto.SALT_LEN);
            off += Crypto.SALT_LEN;
            c.nonce = Arrays.copyOfRange(raw, off, off + Crypto.NONCE_LEN);
            off += Crypto.NONCE_LEN;

            if (Integer.compareUnsigned(c.iterations, Crypto.MIN_ITERATIONS) < 0) {
                throw new StegoException(StegoException.Kind.CORRUPT,
                        "Refusing this payload: its iteration count is below the safe minimum.");
            }
        } else if (c.kdf != KDF_NONE) {
            throw StegoException.badMagic();
        }

        if (raw.length < off + LEN_FIELD) {
            throw StegoException.truncated();
        }
        int n = ByteBuffer.wrap(raw, off, 4).getInt();
        off += LEN_FIELD;

        if (n < 0 || n > MAX_PAYLOAD) {
            throw new StegoException(StegoException.Kind.CORRUPT,
                    "The declared payload size is out of range.");
        }
        if (raw.length - off < n) {
            throw StegoException.truncated();
        }

        c.header = Arrays.copyOfRange(raw, 0, off);
        c.payload = Arrays.copyOfRange(raw, off, off + n);
        return c;
    }

    /** Builds bodyLen || body || padding. */
    public static byte[] packInner(byte[] body, boolean pad) {
        int total = LEN_FIELD + body.length;
        if (pad) {
            total = (total + PAD_BLOCK - 1) / PAD_BLOCK * PAD_BLOCK;
        }

        byte[] inner = new byte[total];
        inner[0] = (byte) (body.length >>> 24);
        inner[1] = (byte) (body.length >>> 16);
        inner[2] = (byte) (body.length >>> 8);
        inner[3] = (byte) body.length;
        System.arraycopy(body, 0, inner, LEN_FIELD, body.length);

        int padLen = total - LEN_FIELD - body.length;
        if (padLen > 0) {
            byte[] noise = Crypto.random(padLen);
            System.arraycopy(noise, 0, inner, LEN_FIELD + body.length, padLen);
        }
        return inner;
    }

    /** Reverses {@link #packInner} and drops the padding. */
    public static byte[] unpackInner(byte[] inner) throws StegoException {
        if (inner.length < LEN_FIELD) {
            throw StegoException.truncated();
        }
        int n = ByteBuffer.wrap(inner, 0, 4).getInt();
        if (n < 0 || n > inner.length - LEN_FIELD) {
            throw StegoException.truncated();
        }
        return Arrays.copyOfRange(inner, LEN_FIELD, LEN_FIELD + n);
    }

    static boolean startsWith(byte[] data, byte[] prefix) {
        if (data.length < prefix.length) {
            return false;
        }
        for (int i = 0; i < prefix.length; i++) {
            if (data[i] != prefix[i]) {
                return false;
            }
        }
        return true;
    }
}
